File: hs/Mod_addKAry.hs
module Mod_addKAry ( addKAry ) where
import Prelude hiding (and, or) 
import Debug.Trace
import Splib
import Types


trace' :: String -> a -> a
trace' str x = (trace str x) 
--trace' str x = x

data Data_addKAry  = Data_addKAry  { 
  base :: Int,
  list1 :: ListInt,
  list2 :: ListInt,
  carry :: Int,
  sum :: Int,
  res :: ListInt 
}

--SIGNATURES
addKAry :: Int -> ListInt -> ListInt -> ListInt
o_init :: Data_addKAry -> ListInt
o_next :: Data_addKAry -> ListInt
o_l1add :: Data_addKAry -> ListInt
o_l2add :: Data_addKAry -> ListInt
o_carry :: Data_addKAry -> ListInt
o_end :: Data_addKAry -> ListInt
o_calc :: Data_addKAry -> ListInt
o_calc_both :: Data_addKAry -> ListInt
o_calc_l1 :: Data_addKAry -> ListInt
o_calc_l2 :: Data_addKAry -> ListInt
l1_empty :: Data_addKAry -> Bool
l2_empty :: Data_addKAry -> Bool
or_empty :: Data_addKAry -> Bool
and_empty :: Data_addKAry -> Bool
carry_set :: Data_addKAry -> Bool

--DEFINITIONS
addKAry base list1 list2 = (o_init (Data_addKAry base list1 list2 undefined undefined undefined))

l1_empty (Data_addKAry base list1 list2 carry sum res) = 
  (empty list1)
l2_empty (Data_addKAry base list1 list2 carry sum res) = 
  (empty list2)
or_empty (Data_addKAry base list1 list2 carry sum res) = 
  (sp_or (empty list1) (empty list2))
and_empty (Data_addKAry base list1 list2 carry sum res) = 
  (sp_and (empty list1) (empty list2))
carry_set (Data_addKAry base list1 list2 carry sum res) = 
  (sp_eq carry 1)

o_init (Data_addKAry base list1 list2 carry sum res) = (trace' "o_init" flow)
  where
    res' = []
    carry' = 0
    base' = base
    list1' = list1
    list2' = list2
    sum' = sum
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      (if' (and_empty data')
        (o_end data')
        (o_calc data')
      )

o_next (Data_addKAry base list1 list2 carry sum res) = (trace' "o_next" flow)
  where
    carry' = (sp_div sum base)
    res' = (prepend res (sp_mod sum base))
    list1' = (removeLast list1)
    list2' = (removeLast list2)
    base' = base
    sum' = sum
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      (o_calc data')

o_l1add (Data_addKAry base list1 list2 carry sum res) = (trace' "o_l1add" flow)
  where
    carry' = (sp_div sum base)
    res' = (prepend res (sp_mod sum base))
    list1' = (removeLast list1)
    base' = base
    list2' = list2
    sum' = sum
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      (if' (l1_empty data')
        (o_carry data')
        (o_calc data')
      )

o_l2add (Data_addKAry base list1 list2 carry sum res) = (trace' "o_l2add" flow)
  where
    carry' = (sp_div sum base)
    res' = (prepend res (sp_mod sum base))
    list2' = (removeLast list2)
    base' = base
    list1' = list1
    sum' = sum
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      (if' (l2_empty data')
        (o_carry data')
        (o_calc data')
      )

o_carry (Data_addKAry base list1 list2 carry sum res) = (trace' "o_carry" flow)
  where
    res' = (sp_if (sp_eq carry 1) (prepend res carry) res)
    base' = base
    list1' = list1
    list2' = list2
    carry' = carry
    sum' = sum
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      (o_end data')

o_end (Data_addKAry base list1 list2 carry sum res) = (trace' "o_end" flow)
  where
    base' = base
    list1' = list1
    list2' = list2
    carry' = carry
    sum' = sum
    res' = res
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      res'

o_calc (Data_addKAry base list1 list2 carry sum res) = (trace' "o_calc" flow)
  where
    base' = base
    list1' = list1
    list2' = list2
    carry' = carry
    sum' = sum
    res' = res
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      (if' (and_empty data')
        (o_carry data')
        (if' (or_empty data')
          (if' (l1_empty data')
            (o_calc_l2 data')
            (o_calc_l1 data')
          )
          (o_calc_both data')
        )
      )

o_calc_both (Data_addKAry base list1 list2 carry sum res) = (trace' "o_calc_both" flow)
  where
    sum' = (sp_add (last list1) (sp_add (last list2) carry))
    base' = base
    list1' = list1
    list2' = list2
    carry' = carry
    res' = res
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      (o_next data')

o_calc_l1 (Data_addKAry base list1 list2 carry sum res) = (trace' "o_calc_l1" flow)
  where
    sum' = (sp_add (last list1) carry)
    base' = base
    list1' = list1
    list2' = list2
    carry' = carry
    res' = res
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      (o_l1add data')

o_calc_l2 (Data_addKAry base list1 list2 carry sum res) = (trace' "o_calc_l2" flow)
  where
    sum' = (sp_add (last list2) carry)
    base' = base
    list1' = list1
    list2' = list2
    carry' = carry
    res' = res
    data' = (Data_addKAry base' list1' list2' carry' sum' res')
    flow = 
      (o_l2add data')

File: cpp/Mod_addKAry.cpp
#pragma once


ListInt addKAry(Int base, ListInt list1, ListInt list2);
ListInt addKAry_debug(Int base, ListInt list1, ListInt list2);

class addKAry_class {
    
    //VARIABLES        
    Int base;
    Int base_spnew;
    ListInt list1;
    ListInt list1_spnew;
    ListInt list2;
    ListInt list2_spnew;
    Int carry;
    Int carry_spnew;
    Int sum;
    Int sum_spnew;
    ListInt res;
    ListInt res_spnew;
    bool debug_spv;
    
    //PREDICATES
    bool l1_empty() {
        return empty(list1);
    }

    bool l2_empty() {
        return empty(list2);
    }

    bool or_empty() {
        return sp_or(empty(list1), empty(list2));
    }

    bool and_empty() {
        return sp_and(empty(list1), empty(list2));
    }

    bool carry_set() {
        return sp_eq(carry, 1);
    }

    //OPERATIONS
    void o_init() {                    
        res_spnew = {};
        carry_spnew = 0;
        
        res = res_spnew;
        carry = carry_spnew;       

        if(debug_spv) {
            cout << "o_init" << " ; "  
             << "" << ";"  << "" << ";"  << "" << ";"  << carry << ";"  << "" << ";"  << res  
            << endl;   
        }

        if(and_empty()) {
            o_end();
        } else {
            o_calc();
        }
    }

    void o_next() {                    
        carry_spnew = sp_div(sum, base);
        res_spnew = prepend(res, sp_mod(sum, base));
        list1_spnew = removeLast(list1);
        list2_spnew = removeLast(list2);
        
        carry = carry_spnew;
        res = res_spnew;
        list1 = list1_spnew;
        list2 = list2_spnew;       

        if(debug_spv) {
            cout << "o_next" << " ; "  
             << "" << ";"  << list1 << ";"  << list2 << ";"  << carry << ";"  << "" << ";"  << res  
            << endl;   
        }

        o_calc();
    }

    void o_l1add() {                    
        carry_spnew = sp_div(sum, base);
        sum_spnew = sp_mod(sum, base);
        res_spnew = prepend(res, sum);
        list1_spnew = removeLast(list1);
        
        carry = carry_spnew;
        sum = sum_spnew;
        res = res_spnew;
        list1 = list1_spnew;       

        if(debug_spv) {
            cout << "o_l1add" << " ; "  
             << "" << ";"  << list1 << ";"  << "" << ";"  << carry << ";"  << sum << ";"  << res  
            << endl;   
        }

        if(l1_empty()) {
            if(carry_set()) {
                o_carry();
            } else {
                o_end();
            }
        } else {
            o_calc();
        }
    }

    void o_l2add() {                    
        carry_spnew = sp_div(sum, base);
        sum_spnew = sp_mod(sum, base);
        res_spnew = prepend(res, sum);
        list2_spnew = removeLast(list2);
        
        carry = carry_spnew;
        sum = sum_spnew;
        res = res_spnew;
        list2 = list2_spnew;       

        if(debug_spv) {
            cout << "o_l2add" << " ; "  
             << "" << ";"  << "" << ";"  << list2 << ";"  << carry << ";"  << sum << ";"  << res  
            << endl;   
        }

        if(l2_empty()) {
            if(carry_set()) {
                o_carry();
            } else {
                o_end();
            }
        } else {
            o_calc();
        }
    }

    void o_carry() {                    
        res_spnew = prepend(res, 1);
        
        res = res_spnew;       

        if(debug_spv) {
            cout << "o_carry" << " ; "  
             << "" << ";"  << "" << ";"  << "" << ";"  << "" << ";"  << "" << ";"  << res  
            << endl;   
        }

        o_end();
    }

    void o_end() {                    

        
       

        if(debug_spv) {
            cout << "o_end" << " ; "  
             << "" << ";"  << "" << ";"  << "" << ";"  << "" << ";"  << "" << ";"  << ""  
            << endl;   
        }

        return;
    }

    void o_calc() {                    

        
       

        if(debug_spv) {
            cout << "o_calc" << " ; "  
             << "" << ";"  << "" << ";"  << "" << ";"  << "" << ";"  << "" << ";"  << ""  
            << endl;   
        }

        if(or_empty()) {
            if(l1_empty()) {
                o_calc_l2();
            } else {
                o_calc_l1();
            }
        } else {
            o_calc_both();
        }
    }

    void o_calc_both() {                    
        sum_spnew = sp_add(last(list1), sp_add(last(list2), carry));
        
        sum = sum_spnew;       

        if(debug_spv) {
            cout << "o_calc_both" << " ; "  
             << "" << ";"  << "" << ";"  << "" << ";"  << "" << ";"  << sum << ";"  << ""  
            << endl;   
        }

        o_next();
    }

    void o_calc_l1() {                    
        sum_spnew = sp_add(last(list1), carry);
        
        sum = sum_spnew;       

        if(debug_spv) {
            cout << "o_calc_l1" << " ; "  
             << "" << ";"  << "" << ";"  << "" << ";"  << "" << ";"  << sum << ";"  << ""  
            << endl;   
        }

        o_l1add();
    }

    void o_calc_l2() {                    
        sum_spnew = sp_add(last(list2), carry);
        
        sum = sum_spnew;       

        if(debug_spv) {
            cout << "o_calc_l2" << " ; "  
             << "" << ";"  << "" << ";"  << "" << ";"  << "" << ";"  << sum << ";"  << ""  
            << endl;   
        }

        o_l2add();
    }
    
    public:
    addKAry_class(Int constr_base, ListInt constr_list1, ListInt constr_list2) : base(constr_base), list1(constr_list1), list2(constr_list2)   {  }
        
    ListInt execute(bool debug) {           
        debug_spv = debug;
        
        if(debug_spv) {
            cout << "addKAry" << ";"  
            << "base"<< "; " << "list1"<< "; " << "list2"<< "; " << "carry"<< "; " << "sum"<< "; " << "res" 
            << endl;             
            
            cout << "" << "; "  
             << base << ";"  << list1 << ";"  << list2 << ";"  << "" << ";"  << "" << ";"  << ""  
            << endl;   
        }
        
        o_init();           
        return res;
    }
        
    
};

ListInt addKAry(Int base, ListInt list1, ListInt list2) {
    addKAry_class obj(base, list1, list2);  
    return obj.execute(false);
}


ListInt addKAry_debug(Int base, ListInt list1, ListInt list2) {
    addKAry_class obj(base, list1, list2);  
    return obj.execute(true);
}